### PR TITLE
Improve the autogenerated note in RST docs for class reference

### PR DESCRIPTION
**Pull request by YuriSizov**
This note hasn't been changed in quite some time, and I've noticed we could actually refer to the actual file path, so potential contributors don't need to look through all the modules.

Also some phrasing was a bit awkward. It now looks something like this:

```rst

:github_url: hide

.. DO NOT EDIT THIS FILE!!!
.. Generated automatically from Godot engine sources.
.. Generator: https://github.com/godotengine/godot/tree/master/doc/tools/make_rst.py.
.. XML source: https://github.com/godotengine/godot/tree/master/modules/gdscript/doc_classes/@GDScript.xml.

.. _class_@GDScript:

@GDScript
=========

:github_url: hide

.. DO NOT EDIT THIS FILE!!!
.. Generated automatically from Godot engine sources.
.. Generator: https://github.com/godotengine/godot/tree/master/doc/tools/make_rst.py.
.. XML source: https://github.com/godotengine/godot/tree/master/doc/classes/@GlobalScope.xml.

.. _class_@GlobalScope:

@GlobalScope
============

```

-----

I've noticed that we disable GitHub URL on these pages, but we could instead use the generated URL for the source XML and keep the button visible. Just food for thought.

### CODE REVIEW

**Comment by akien-mga** [comment]
I would provide a direct URL to the XML over a direct URL to `make_rst.py`.

**Comment by YuriSizov** [comment]
> that being said it could end up becoming a 404 when we rename classes so maybe it's not a good idea.

Yeah, that would make sense, if we could pin the commit this was generated with. Any suggestions on how we could do that? Considering this is something that you do manually.

PS. Should we set up a job to sync up the class reference at some point?

**Comment by akien-mga** [comment]
You could import `version.py` and use the matching docs branch maybe?

**Comment by YuriSizov** [comment]
> You could import `version.py` and use the matching docs branch maybe?

Well, I figured if we're importing `version.py` to fetch a hard-configured value, then we can just as well add a new value for the reference Git branch. So that's what I did.

If you're gonna cherrypick it, don't forget to change the branch in `version.py` (btw, `docs` only exists in 3.5, but not before; maybe worth backporting it as well to earlier branches, for parity).

**Comment by akien-mga** [comment]
I'm not fond of adding yet another variable to `version.py` for such a niche use case.

Just use `docs` and special case the `latest -> master` conversion in `make_rst.py`.

**Comment by YuriSizov** [comment]
Well, we could use it for something else in the future too. I didn't find any GitHub links in the editor before making this change, to justify it better, but it could still be handy in future 🙃 

I just figured, when you need to change this, you need to go to `version.py` and edit it already, to link to different docs branch. Wouldn't be much effort to edit two lines instead of just one. And it would only be done once per starting a new branch, and if we ever move class ref updates to CI, we could pass a commit hash this way too. And it would mean less weird magic in `make_rst.py`. Of course, this can be changed at any point in the future.

But if you insist, I'll change it in a bit.

**Comment by akien-mga** [comment]
`version.py` is for stuff that will end up in `core/version_generated.gen.h`. That's not the case for this new field which is only needed for a tool not needed for Godot itself, so I wouldn't add this there.

If you really don't want to special case it, then you can parse the Git head instead, extracting relevant logic from `methods.update_version`. But IMO that's not necessary and goes against our "keep solutions local" mantra.

**Comment by YuriSizov** [comment]
> Adding a new field to one of the most core files of the project just for the sake of a generated comment that users shouldn't ever see is really not something I'm keen on doing :)

Maybe then we could add a field in `make_rst.py` itself, so it's still configurable, just directly there? Shouldn't affect cherrypicks for other code changes in future.

**Comment by akien-mga** [comment]
Why not, though `version.docs` *is* really exactly that already with the caveat that `latest` should be mapped to `master`. For `3.4`, `3.5`, `3.6`, `4.0`, etc. it will be correct.

**Comment by YuriSizov** [comment]
Did it with `version.docs`, but do keep in mind that 3.4 doesn't have it.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: doc/tools/make_rst.py [hunk 1]
```diff
@@ -587,12 +592,26 @@ def make_rst_class(class_def, state, dry_run, output_dir):  # type: (ClassDef, S
...
-    f.write(".. The source is found in doc/classes or modules/<name>/doc_classes.\n\n")
 
+    # Warn contributors not to edit this file directly.
+    # Also provide links to the source files for reference.
+
+    git_branch = "master"
+    if hasattr(version, "docs") and version.docs != "latest":
+        git_branch = version.docs
+
+    source_xml_path = os.path.normpath(os.path.join("doc", class_def.filepath)).replace("\\", "/")
```
Comment (akien-mga) [inline doc/tools/make_rst.py:605]: Doesn't work on Linux:
```
.. XML source: https://github.com/godotengine/godot/tree/master//home/akien/Projects/godot/godot.git/doc/classes/AESContext.xml.
```

Comment (YuriSizov) [reply doc/tools/make_rst.py:605]: Huh, didn't expect the path not to be relative. Let me investigate it a bit more.

Comment (akien-mga) [reply doc/tools/make_rst.py:605]: Try `os.path.relpath` maybe?

Comment (akien-mga) [reply doc/tools/make_rst.py:605]: This works for me:
```diff
diff --git a/doc/tools/make_rst.py b/doc/tools/make_rst.py
index a56e4b338e..12d153a7aa 100755
--- a/doc/tools/make_rst.py
+++ b/doc/tools/make_rst.py
@@ -602,7 +602,7 @@ def make_rst_class(class_def, state, dry_run, output_dir):  # type: (ClassDef, S
     if hasattr(version, "docs") and version.docs != "latest":
         git_branch = version.docs
 
-    source_xml_path = os.path.normpath(os.path.join("doc", class_def.filepath)).replace("\\", "/")
+    source_xml_path = os.path.relpath(class_def.filepath, script_directory + "/../..").replace("\\", "/")
     source_github_url = "https://github.com/godotengine/godot/tree/{}/{}".format(git_branch, source_xml_path)
     generator_github_url = "https://github.com/godotengine/godot/tree/{}/doc/tools/make_rst.py".format(git_branch)
 
```

Comment (akien-mga) [reply doc/tools/make_rst.py:605]: Fix in https://github.com/godotengine/godot/pull/63530

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.

**Pull request merged**